`IndexedSeq#head` now throws `NoSuchElementException` (not `IndexOutOfBoundsException`)

som-snytt opened the pull request:
Fixes scala/bug#12782

Attempts to use `collectionClassName` for a helpful exception message, falls back to `toString`, which should be fine if the collection is in fact empty. (Extra parens at worst.)

~Avoids~ Relishes `isEmpty` in the happy path.

som-snytt committed:
Translate outbound IndexOutOfBounds in IndexedSeq.head

sjrd commented on src/library/scala/collection/IndexedSeq.scala:
```diff
@@ -91,11 +91,29 @@ trait IndexedSeqOps[+A, +CC[_], +C] extends Any with SeqOps[A, CC, C] { self =>
+  override def head: A =
+    try apply(0)
+    catch {
+      case e: IndexOutOfBoundsException =>
```
This won't work on Scala.js. Array index out of bounds is undefined behavior over there. Besides, using a `try` for the happy path is not quite as free as on the JVM.

It seems to me that an explicit `isEmpty` test would be more appropriate. It's what we have in other collections I believe.

som-snytt replied:
I waffled on `isEmpty` because of hidden costs, but maybe that is paranoid for an indexed thing. Looking at it now, I see the PR from last year was that `IterableOnce#isEmpty` now uses `knownSize` first and `iterator.hasNext` second. I haven't looked whether indexed implies size is known, but it seems likely. They have to override something useful if they want a useful collection.

Edit: I haven't internalized that `IndexedSeq` means "efficient `apply` and `length`" and `knownSize` is just `length`. Also `SeqOps#isEmpty` is `lengthCompare(0)`, so probably `isEmpty` is guaranteed to be fine.

`IndexOutOfBoundsException` doesn't imply `ArrayIndexOutOfBounds`; this behavior is only for collections which don't override `head`. I didn't quite do a survey but picked `ArrayDeque` as an example, and it happens to do an explicit bounds check.

"Ideal" would be a comprehensive unit and benchmark test for performance and coverage (which collections or wrappers fail to do something specific for `head` or `isEmpty`?). I said I didn't benchmark last year, so I'll try that.

Also meant to add, I figured out how to invoke
```
sbt:root> junit/testOnly scala.collection.IndexedTestImpl.ArrayBufferTest
```
but the types aren't constrained where you can toss in a test of `xs.head`, so I gave up quickly. Also in progress is a tweak or update to `collections-laws`.

som-snytt replied:
Just to document that the first attempt was `new NSEE(ioobe.getMessage, ioobe.getCause)` but actually the new exception should have had the `NSEE` as the cause.

som-snytt committed:
IndexedSeq.head uses isEmpty

Seq.isEmpty is lengthCompare and IndexedSeq
has efficient length aka knownSize. headOption
already uses isEmpty.

som-snytt commented:
I caved to sjrd's reasonable demands.

sjrd reviewed (approved):
Thanks. This looks good to me. :)

odd reviewed (approved):
LGTM

kitbellew commented on src/library/scala/collection/IndexedSeq.scala:
```diff
@@ -91,11 +91,25 @@ trait IndexedSeqOps[+A, +CC[_], +C] extends Any with SeqOps[A, CC, C] { self =>
   override def slice(from: Int, until: Int): C = fromSpecific(new IndexedSeqView.Slice(this, from, until))
 
-  override def head: A = apply(0)
+  override def head: A =
```
wouldn't this now lead to repeated checks for `isEmpty` in `{head,last}Option`? should there be an `@inline def headImpl: A = apply(0)` and used in both `head` and `headOption`?

som-snytt replied:
My comment on the first commit is that head is fast. `headOption` is allocating.

SethTisue commented:
Release-noting since there are compatibility implications here — for those catching exceptions, and for those extending `IndexedSeq`, as we're seeing over at https://github.com/scala/community-build/issues/1680
